### PR TITLE
[x] add initial support for automatic workspace-hack management

### PR DESCRIPTION
**Pull request by sunshowers**
# Automatic workspace-hack management

tl;dr: **faster local and CI builds** through automating feature unification in Diem Core; run `cargo x generate-workspace-hack` to regenerate contents.

## What are workspace-hack packages?

Diem, like many other large Rust workspaces, tends to have trouble with slow rebuilds because of feature unification issues. The way we’ve addressed these issues is to introduce a [*workspace-hack* package](https://github.com/diem/diem/blob/master/common/workspace-hack/Cargo.toml), which every other crate depends upon and which ensures that features are unified for the listed third-party dependencies.

For more details about workspace-hack packages, see [this documentation](https://facebookincubator.github.io/cargo-guppy/rustdoc/hakari/).

## What’s changing?

So far we, and other projects like us, have maintained these workspace-hack packages by hand. However, I’ve been working on a tool called [hakari](https://github.com/facebookincubator/cargo-guppy/tree/main/tools/hakari), which leverages [guppy](https://crates.io/crates/guppy) to automate management of these packages.

Once this PR lands:

* the contents of the diem-workspace-hack package will be automatically managed
* there will be a new lint which ensures that the contents are up-to-date
* if they’re out of date then the lint will tell you to run a new command, `cargo x generate-workspace-hack`, which will regenerate the contents

## What are the benefits?

The main invariant hakari upholds is that no matter which package or set of packages you build in the workspace, **all third-party dependencies will be built with exactly the same feature set**. In our tests, this improves caching and reduces local + CI build times quite significantly.

The following figures were taken on a MacBook Pro 16“ with a Intel(R) Core(TM) i9-9980HK CPU (8c/16t) running macOS 10.15.7. The command run was:

`cargo clean && time cargo x --help && time cargo xbuild -p diem-vm && time cargo xbuild -p diem-types && time cargo xbuild -p diem-key-manager && time cargo xbuild -p execution-correctness && time cargo xbuild -p diem-node`

|Command	|Before, wall (s)	|After, wall (s)	|Difference (s, negative means after is faster)	|% difference (from before)	|Cumulative difference (s)	|Comment	|
|---	|---	|---	|---	|---	|---	|---	|
|cargo x —help	|45.23	|45.58	|0.35	|0.77%	|0.35	|Negligible impact	|
|cargo xbuild -p diem-vm	|53.76	|77.24	|23.48	|43.68%	|23.83	|First build takes longer because more dependencies have to be unified	|
|cargo xbuild -p diem-types	|13.35	|0.91	|-12.44	|-93.18%	|11.39	|Core package, build becomes a no-op	|
|cargo xbuild -p diem-key-manager	|52.42	|45.58	|-6.84	|-13.05%	|4.55	|Requires more third-party deps	|
|cargo xbuild -p execution-correctness	|20.07	|14.02	|-6.05	|-30.14%	|-1.50	|	|
|cargo xbuild -p diem-node	|171.16	|146.97	|-24.19	|-14.13%	|-25.69	|Builds most of Diem, still much faster	|

For the raw data, including results from a Linux desktop, see [this gist](https://gist.github.com/sunshowers/9852d3fb0e46a7a468886a9084015405).

## What are the tradeoffs?

1. Merge conflicts are may arise in the workspace-hack’s Cargo.toml; this file is configured to not leave conflict markers, and you can run `cargo x generate-workspace-hack` to resolve them locally. Unfortunately we don’t yet have the ability to resolve them automatically on GitHub—doing so may be possible in some circumstances but is quite complicated and will require prioritization.

2. If we left the workspace-hack in place for our release artifacts (Docker images), we’d be including a ton more code in the TCB than necessary. This is undesirable, so we now **disable the workspace hack** while performing the Docker build. This means that our effective TCB will become smaller thanks to this PR (the dependency numbers we currently report were already with the workspace-hack disabled, but our build process didn’t actually do that before this PR.
    
    The tradeoff is that this introduces a divergence between developer and production builds — based on the factors involved, we believe that this is the best choice. To replicate production builds and disable the workspace-hack locally, run `cargo x generate-workspace-hack --mode disable`. To re-enable the workspace-hack, simply run `cargo x generate-workspace-hack` again.

### CODE REVIEW

**Comment by sunshowers** [comment]
Ahh, looks like the unification's causing openssl to be built, which we don't support :( going to think about this more.

**Comment by sunshowers** [comment]
So I think my biggest concern here is that we don't ever build our release binaries with the workspace-hack package disabled. I think that would be very useful to work on...

**Comment by bmwill** [comment]
> Ahh, looks like the unification's causing openssl to be built, which we don't support :( going to think about this more.

Has this been solved? As in is this ready for a thorough review now?

**Comment by sunshowers** [comment]
Yeah it's solved, and ready for review now

**Review by bmwill — APPROVED** [review]
Looks good! I'm excited to see how well this reduces rebuilds!

File: Cargo.toml [hunk 1]
```diff
@@ -199,3 +199,10 @@ debug = true
 # Build guppy in opt mode so that x commands are faster.
 [profile.dev.package.guppy]
 opt-level = 3
+
+[patch.crates-io]
+guppy = { git = "https://github.com/facebookincubator/cargo-guppy", branch = "main" }
+target-spec = { git = "https://github.com/facebookincubator/cargo-guppy", branch = "main" }
+guppy-summaries = { git = "https://github.com/facebookincubator/cargo-guppy", branch = "main" }
+determinator = { git = "https://github.com/facebookincubator/cargo-guppy", branch = "main" }
+hakari = { git = "https://github.com/facebookincubator/cargo-guppy", branch = "main" }
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply Cargo.toml:208]: At the moment yeah, I plan to do releases soon (probably will do releases + update before landing this)

File: common/workspace-hack/Cargo.toml [hunk 2]
```diff
@@ -9,22 +9,183 @@ license = "Apache-2.0"
...
-num-integer = { version = "0.1.44", features = ["i128", "std"] }
-num-traits = { version = "0.2.14", features = ["i128", "std"] }
-once_cell = "1.5.2"
-petgraph = { version = "0.5.1", features = ["graphmap", "matrix_graph", "stable_graph"] }
-serde = { version = "1.0.120", features = ["derive", "rc"] }
-sha-1 = { version = "0.9.2", features = ["std"] }
-subtle = { version = "2.4.0", features = ["i128", "std"] }
+### BEGIN HAKARI SECTION
+[target.x86_64-unknown-linux-gnu.dependencies]
+base64 = { version = "0.13.0", default-features = true, features = ["alloc", "default", "std"] }
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply common/workspace-hack/Cargo.toml:14]: Yeah, the auto-generation code always outputs default-features at the moment. I guess I'll remove it.

File: devtools/x/src/generate_workspace_hack.rs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+            match args.mode {
+                WorkspaceHackMode::Write => {
+                    // Write out the contents to the TOML file.
+                    existing_toml.write_to_file(&new_toml)?;
+                }
+                WorkspaceHackMode::Diff => {
+                    let patch = existing_toml.diff_toml(&new_toml);
+                    // TODO: add global coloring options to x
+                    let formatter = hakari::diffy::PatchFormatter::new().with_color();
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply devtools/x/src/generate_workspace_hack.rs:72]: yeah :D it's so great!

File: common/workspace-hack/Cargo.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (sunshowers) [reply common/workspace-hack/Cargo.toml]: Done, we no longer output `default-features = true`

**Comment by sunshowers** [comment]
Made a couple minor updates and one major update per discussion with @jnaulty: disable the workspace-hack for release builds.

File: Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (bmwill) [reply Cargo.toml:208]: Still planning on publishing before landing or will that come at a later point in time?

[the remark this answers is not in the document]
Comment (sunshowers) [reply Cargo.toml:208]: Will still do it before landing, planning to make the releases today or tomorrow. There were a surprising number of API changes I needed to make to get this working.

[the remark this answers is not in the document]
Comment (sunshowers) [reply Cargo.toml]: Released new versions to crates.io.
